test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala

```diff
@@ -0,0 +1,329 @@
+package org.scalajs.testsuite.javalib.time
+
+import java.time._
+import java.time.temporal._
+
+import org.junit.Test
+import org.junit.Assert._
+import org.scalajs.testsuite.utils.AssertThrows._
+
+/** Created by alonsodomin on 24/12/2015. */
+class YearTest extends TemporalTest[Year] {
+  import Year._
+  import ChronoUnit._
+  import ChronoField._
+  import DateTimeTestUtil._
+
+  val min = Year.of(MIN_VALUE)
+  val max = Year.of(MAX_VALUE)
+  val leapYear = Year.of(2016)
+  val nonLeapYear = Year.of(2015)
+  val lastBCYear = Year.of(0)
+  val firstACYear = Year.of(1)
+  val negativeYear = Year.of(-100)
+
+  val samples = Seq(min, max, leapYear, nonLeapYear, lastBCYear, firstACYear, negativeYear)
+
+  override def isSupported(unit: ChronoUnit): Boolean =
+    unit == YEARS || unit == DECADES || unit == CENTURIES || unit == MILLENNIA || unit == ERAS
+
+  override def isSupported(field: ChronoField): Boolean =
+    field == YEAR_OF_ERA || field == YEAR || field == ERA
+
+  override def expectedRangeFor(accessor: Year, field: TemporalField): ValueRange = {
+    field match {
+      case YEAR_OF_ERA =>
+        if (accessor.getValue() <= 0) ValueRange.of(1, MAX_VALUE + 1)
+        else ValueRange.of(1, MAX_VALUE)
+
+      case _ =>
+        super.expectedRangeFor(accessor, field)
+    }
+  }
+
+  @Test def getLong(): Unit = {
+    for {
+      t <- samples
+      field <- ChronoField.values()
+    } {
+      if (isSupported(field)) {
+        val expected = (field match {
+          case YEAR_OF_ERA => if (t.getValue < 1) 1 - t.getValue else t.getValue
+          case YEAR        => t.getValue
+          case ERA         => if (t.getValue < 1) 0 else 1
+          case _           =>
+            // Should not happen
+            throw new UnsupportedTemporalTypeException(s"Unsupported field: $field")
+        }).toLong
+
+        assertEquals(expected, t.getLong(field))
+      } else {
+        expectThrows(classOf[UnsupportedTemporalTypeException], t.getLong(field))
+      }
+    }
+
+    assertEquals(1L, lastBCYear.getLong(YEAR_OF_ERA))
+    assertEquals(0L, lastBCYear.getLong(YEAR))
+    assertEquals(0L, lastBCYear.getLong(ERA))
+
+    assertEquals(101L, negativeYear.getLong(YEAR_OF_ERA))
+    assertEquals(-100L, negativeYear.getLong(YEAR))
+    assertEquals(0L, negativeYear.getLong(ERA))
+  }
+
+  @Test def length(): Unit = {
+    assertEquals(366, leapYear.length())
+    assertEquals(365, nonLeapYear.length())
+  }
+
+  @Test def `with`(): Unit = {
+    assertEquals(lastBCYear, lastBCYear.`with`(YEAR, 0))
+    assertEquals(Year.of(0), lastBCYear.`with`(YEAR_OF_ERA, 1))
```

Comment:
Add also the case where `with(YEAR_OF_ERA, 0)`. Do the same for the other years bellow.